### PR TITLE
#130 Deploy web on api server

### PR DESCRIPTION
**Pull request by SungHwan-Park**

### CODE REVIEW

**Review by suzalflueck — APPROVED** [review]

File: app/.pipeline/lib/build.js [hunk 1]
```diff
@@ -11,15 +11,15 @@ module.exports = (settings)=>{
 
   const templatesLocalBaseUrl =oc.toFileUrl(path.resolve(__dirname, '../../openshift'))
 
-  objects.push(...oc.processDeploymentTemplate(`${templatesLocalBaseUrl}/client/bc.yaml`, {
```
Comment (SungHwan-Park) [inline app/.pipeline/lib/build.js:14]: skip building web image

File: app/.pipeline/lib/deploy.js [hunk 2]
```diff
@@ -17,16 +17,16 @@ module.exports = (settings)=>{
   mongoCredential['pass'] = Buffer.from(mongoSecret.data['database-password'], 'base64').toString('utf-8');
 
   const templatesLocalBaseUrl =oc.toFileUrl(path.resolve(__dirname, '../../openshift'))
-
-  objects.push(...oc.processDeploymentTemplate(`${templatesLocalBaseUrl}/client/dc.yaml`, {
```
Comment (SungHwan-Park) [inline app/.pipeline/lib/deploy.js:21]: skip deploying web pods

File: app/client/server.js [hunk 3]
```diff
@@ -1,7 +1,16 @@
 const express = require('express');
+const path = require('path');
 
 const app = express();
-app.use(express.static('./build'));
+const buildPath = path.join(__dirname, './build');
```
Comment (SungHwan-Park) [inline app/client/server.js:5]: these codes are just for running web on its own pod and on nodejs.
if we deploy web on api or on other types of web server like nginx, also we don't need this file.

File: app/client/src/views/Dashboard.tsx [hunk 4]
```diff
@@ -46,67 +46,67 @@ const Dashboard: React.FC = () => {
 
   const [profile, setProfile] = useState<any>([]);
   const [tableView, setTableView] = useState(true);
-
-  useEffect(() => {
```
Comment (SungHwan-Park) [inline app/client/src/views/Dashboard.tsx:50]: most of useEffect should be disabled until we integrate web with our api.

File: app/openshift/server/bc.yaml [hunk 5]
```diff
@@ -68,7 +68,7 @@ objects:
       resources: {}
       runPolicy: Serial
       source:
-        contextDir: app/server
```
Comment (SungHwan-Park) [inline app/openshift/server/bc.yaml:71]: to deploy web and api together

File: app/package.json [hunk 6]
```diff
@@ -0,0 +1,25 @@
+{
```
Comment (SungHwan-Park) [inline app/package.json:1]: it is for oc to build and deploy. s2i uses `npm install`, `npm build`, and `npm start` on the `contextDir` folder

File: app/server/src/app.ts [hunk 7]
```diff
@@ -95,7 +97,13 @@ class App {
     this.app.use(cors({ origin: config.get('apiUrl'), credentials: true }));
 
     this.app.use(hpp());
-    this.app.use(helmet());
+    this.app.use(
```
Comment (SungHwan-Park) [inline app/server/src/app.ts:100]: I don't exactly understand why, but the page doesn't load without this security setting in production mode.

File: app/server/src/app.ts [hunk 8]
```diff
@@ -132,6 +140,19 @@ class App {
     routes.forEach(route => {
       this.app.use(`${this.api_root}/${route.resource}`, route.router);
     });
+
+    // serve react pages if it exists
+    const webPath = path.join(__dirname, '../web');
```
Comment (SungHwan-Park) [inline app/server/src/app.ts:145]: api serves web resources as the last route.

**Pull request merged**